types/history/history-tests.ts

```diff
@@ -56,17 +56,19 @@ let input = { value: "" };
     // Extending an existing location object.
     history.push({ pathname: location.pathname, search: '?other=search' });
 
-    if (history.canGo(-1)) {
-        history.go(-1);
-        history.goBack();
-    }
+    // canGo does not exist on History
```

Comment:
It appears to come from `MemoryHistory` in createMemoryHistory.d.ts.  It seems like there's only a problem because of the upcast?

Reply:
You are correct, I hadn't noticed the MemoryHistory type. I've updated it to be Generic and uncommented the tests.

Reply:
Thanks!  Are there any other subtypes of History that ought to be updated at the same time?

Reply:
Currently both createBrowserHistory and createHashHistory both just return objects of type History, although after reviewing the docs this may not be correct.

From the docs, it appears that structurally MemoryHistory extends BrowserHistory extends HashHistory

I see two ways to proceed here, I'm not sure what the DT best practices say to do.
The cleanest would be
interface HashHistory {
    ...
}
interface BrowserHistory extends HashHistory {
    ...
}
interface MemoryHistory extends BrowserHistory {
    ...
}
but that feels wrong because BrowserHistory would depend on HashHistory and MemoryHistory would depend on BrowserHistory, even though the actual source doesn't have those dependencies. 

The other option would be
interface BaseHistory {
    ...
}
and 
interface HashHistory extends BaseHistory {
    // No additional properties
}
interface BrowserHistory extends BaseHistory {
    ...
}
interface MemoryHistory extends BaseHistory {
    ...
}
but this case leads to an empty HashHistory and all of the BrowserHistory properties will be mirrored in MemoryHistory.

Which option do you think is best? I will proceed with that implementation.

Reply:
I don't think the answer is specific to TS or DT.  If a new member gets added to `HashHistory`, do you expect it to also be available in `BrowserHistory` and `MemoryHistory`?  If so, the types should be related.  Otherwise, duplication is fine because the overlap is effectively a coincidence.